### ISSUE TITLE
Generate an error or warning when building under the downloads folder

### ISSUE THREAD

**Issue by dsplaisted** [dotnet/sdk#45342]
Building a project under the downloads folder is not very safe.  This is because of things like Directory.Build.props, which mean that you need to trust all of the parent folders of your source code up to the root.  Most browsers will download files to the downloads folder with no additional confirmation when you click a link, so a malicious link could put a Directory.Build.props file in your downloads folder, and then execute code in your build when you subsequently built a project that had been downloaded (or extracted) under the downloads folder.

We should probably fail outright, or at least warn, when trying to build a project under the Downloads folder.

This probably makes sense to be an MSBuild feature.  Possibly it should be a BuildCheck, but maybe it should just be built into MSBuild by default before it even tries to start the build.

**Comment by Forgind** [comment]
<security label>

**Comment by marcpopMSFT** [comment]
@dotnet/msbuild @dsplaisted we should have a discussion on this one of the best path forward. Should we block? Should we buildcheck/warn? Does this have to be done in msbuild land?

**Comment by YuliiaKovalova** [comment]
@JanKrivanek has already developed a custom build check to track invalid location:
https://github.com/dotutils/BuildChecks/blob/main/src/UntrustedLocationCheck.cs

It's available in this nuget, you can give it a try right away
https://www.nuget.org/packages/DotUtils.BuildChecks/0.0.6-beta

**Comment by marcpopMSFT** [comment]
Ooh, very cool. It's already there. Are there plans to add this to the built-in buildchecks?

**Comment by marcpopMSFT** [comment]
@JanKrivanek thoughts on including this as a built-in eventually?

**Comment by JanKrivanek** [comment]
Detecting Downloads folder is surprisingly complicated - so I'm not sure we want to have it as a default diagnostic, but a built-in buildcheck sounds fine.
Btw. just opening the project in IDE is same dangerous, however less obvious to users.

More general version of this check is "You are importing logic that is outside of the root of your solution/repository/..."

This is just one of many security footguns that MSBuild can bring - we have a due to openly document those (https://github.com/dotnet/msbuild/issues/10739) - so that we can then start pointing users from such new diagnostics to more comprehensive overview, so that we prevent false sense of security by preventing just some of those (we cannot prevent everything that is dangerous as extensibility points are vital part of MSBuild. Over time I'm hoping to collect more tel around usage of various extensibility points and start flipping the more esoteric ones to explicit opt-in mode)

**Comment by marcpopMSFT** [comment]
@JanKrivanek my understanding is that there is already a buildcheck for this. What are the plans to add that included in the built-in ones?

**Comment by JanKrivanek** [comment]
This is not currently planned (other areas of investments are getting precedence in a near future).

But I'll try to find same spare time to copy it to built-in checks - stay tunned :-)

**Comment by marcpopMSFT** [comment]
@JanKrivanek buildchecks run after the build has completed? Daniel points out that this would be too late if there were something malicious going on. Could msbuild startup code potentially block this entirely (unless an env is set) and is that something you'd want to consider?

**Comment by JanKrivanek** [comment]
Yeah I was considering to have a 'panic' mode in the past, but then just left the idea.
Somenoe else bringing this up makes me to at least log it as an issue and have the team triage: https://github.com/dotnet/msbuild/issues/11314

It would be very easy to add and would have potential to bridge the gap. The disadvantage is, that while BuildChecks run during the build (they have the post-hoc replay mode as well - but that's mostly for builds that already happened 'sometime, somewhere' - so understandably not capable to interfere) - they mostly run on top of the logging infra, that is asynchronous to the build execution - so there is still some small window of opportunity for the malitious action. The built-in checks have an ability to load directly in the executing node and interfere directly - so to prevent such window of opportunity, the security checks would need to run in that mode.
